common/src/main/java/org/apache/comet/vector/CometDictionary.java

```diff
@@ -59,121 +46,57 @@ public ValueVector getValueVector() {
 
   public int decodeToInt(int index) {
-    return ints[index];
+    return values.getInt(index);
   }
 
   public long decodeToLong(int index) {
-    return longs[index];
+    return values.getLong(index);
   }
 
   public float decodeToFloat(int index) {
-    return floats[index];
+    return values.getFloat(index);
   }
 
   public double decodeToDouble(int index) {
-    return doubles[index];
+    return values.getDouble(index);
   }
 
   public byte[] decodeToBinary(int index) {
-    return binaries[index].bytes;
-  }
-
-  public UTF8String decodeToUTF8String(int index) {
-    return strings[index];
-  }
-
-  @Override
-  public void close() {
-    values.close();
-  }
-
-  private void initialize() {
     switch (values.getValueVector().getMinorType()) {
-      case BIT:
-        booleans = new boolean[numValues];
-        for (int i = 0; i < numValues; i++) {
-          booleans[i] = values.getBoolean(i);
-        }
-        break;
-      case TINYINT:
-        bytes = new byte[numValues];
-        for (int i = 0; i < numValues; i++) {
-          bytes[i] = values.getByte(i);
-        }
-        break;
-      case SMALLINT:
-        shorts = new short[numValues];
-        for (int i = 0; i < numValues; i++) {
-          shorts[i] = values.getShort(i);
-        }
-        break;
-      case INT:
-      case DATEDAY:
-        ints = new int[numValues];
-        for (int i = 0; i < numValues; i++) {
-          ints[i] = values.getInt(i);
-        }
-        break;
-      case BIGINT:
-      case TIMESTAMPMICRO:
-      case TIMESTAMPMICROTZ:
-        longs = new long[numValues];
-        for (int i = 0; i < numValues; i++) {
-          longs[i] = values.getLong(i);
-        }
-        break;
-      case FLOAT4:
-        floats = new float[numValues];
-        for (int i = 0; i < numValues; i++) {
-          floats[i] = values.getFloat(i);
-        }
-        break;
-      case FLOAT8:
-        doubles = new double[numValues];
-        for (int i = 0; i < numValues; i++) {
-          doubles[i] = values.getDouble(i);
-        }
-        break;
       case VARBINARY:
       case FIXEDSIZEBINARY:
-        binaries = new ByteArrayWrapper[numValues];
-        for (int i = 0; i < numValues; i++) {
-          binaries[i] = new ByteArrayWrapper(values.getBinary(i));
-        }
-        break;
-      case VARCHAR:
-        strings = new UTF8String[numValues];
-        for (int i = 0; i < numValues; i++) {
-          strings[i] = values.getUTF8String(i);
-        }
-        break;
+        return values.getBinary(index);
       case DECIMAL:
-        binaries = new ByteArrayWrapper[numValues];
-        for (int i = 0; i < numValues; i++) {
-          // Need copying here since we re-use byte array for decimal
-          byte[] bytes = new byte[DECIMAL_BYTE_WIDTH];
-          bytes = values.copyBinaryDecimal(i, bytes);
-          binaries[i] = new ByteArrayWrapper(bytes);
-        }
-        break;
+        byte[] bytes = new byte[DECIMAL_BYTE_WIDTH];
```

Comment:
Hmm will this be more expensive? originally we would copy binaries into `ByteArrayWrapper` only once up-front, but now we are doing it for each `decodeToBinary` call?

Reply:
I think for whole stage codgen, the accessing code to values in rows are collapsed as a whole and the value is accessed as a temporary variable in Java code. One value should have only one `decodeToBinary` call.

Although for interpreted execution, each time to access a value in a row, it will call these accessors.

Reply:
We probably can benchmark it. cc @andygrove

Reply:
Btw, except for the cases that only CometScan is triggered, if there is any Comet native operator, these copied dictionary values are not used at all.

Reply:
> I think for whole stage codgen, the accessing code to values in rows are collapsed as a whole and the value is accessed as a temporary variable in Java code. One value should have only one decodeToBinary call.

Hmm I'm not sure. Suppose we have a Comet dictionary vector where indices are 100 1's. In the current approach we'd need to call `decodeToBinary` on index 1 for 100 times? 

> Btw, except for the cases that only CometScan is triggered, if there is any Comet native operator, these copied dictionary values are not used at all. They are only for row-based access by Spark operator.

Yes this is true. 

I'm basically trying to understand whether the changes in this PR is necessary. Is it trying to address some perf regression we saw recently?

Reply:
> I'm basically trying to understand whether the changes in this PR is necessary. Is it trying to address some perf regression we saw recently?

This is motivated by recent regression on CometDictionary after re-initiating CometDictionary after re-importing array, though it is not directly related. It inspires me that this part of copying might be not good for performance.

Maybe we just need to do copying for binary types? For other types like int, long, etc, I don't see the good reason to copy the values actually.

Reply:
Yes, I think should be OK to avoid copying for other types.

Reply:
> We probably can benchmark it. cc @andygrove

I ran the TPC-H benchmark comparing `main` with this PR and did not see any significant difference. I think the changes in this PR make sense, though, and reduce some memory pressure.